_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+
+Wyobraźmy sobie zatem sytuację, w której wszystkie możliwe dane w sposób uporządkowany trafiają do jednego miejsca i mogą być z niego odczytane w dowolnej chwili. Prawda, że cudna koncepcja? Nazywamy ją "single source of truth", czyli, w wolnym tłumaczeniu, jedno źródło prawdy.  Wykorzystanie takiego pomysłu pozwala pisać bardziej przewidywalny i łatwiej testowalny kod, jak również znacząco zwiększa możliwość skalowania aplikacji. W tym wpisie postaram się nieco przybliżyć jedną z implementacji architektury Redux wykorzystującej koncepcję globalnego stanu aplikacji oraz pokażę przykład jej zastosowania w bibliotece NgRx.
+
+### Czy na pewno potrzebujesz zarządzania stanem aplikacji?
+Zanim zaczniemy, chciałbym jeszcze zwrócić uwagę, że nie każda aplikacja jest tak rozbudowana i złożona, żeby potrzebować całego mechanizmu zarządzania stanem. W ustaleniu, czy tak jest, pomaga zasada SHARI zaprezentowana, chociażby, w [oficjalnej dokumentacji NgRx](https://ngrx.io/guide/store/why). Warto sobie odpowiedzieć, czy potrzebujemy stanu, który jest:
+- **S**hared - współdzielony pomiędzy wiele komponentów i serwisów.
+- **H**ydrated - trwały i z możliwością ponownego zasilenia z zewnętrznego źródła jak np. local storage.
+- **A**vailable - dostępny cały czas, niezależnie od sposobu nawigacji po aplikacji, np. podczas przechodzenia i cofania się w aplikacji prezentującej złożony wniosek.
+- **R**etrieved - zdolny do przechowywania danych pochodzących z zewnętrznych źródeł, co pozwala na przykład zapisać wynik żądania HTTP i nie wykonywać go po raz kolejny w celu otrzymania tych samych informacji.
+- **I**mpacted - zdolny do zmieniania się pod wpływem akcji wykonywanych przez komponenty i serwisy.
+
+## Architektura Redux
+No dobrze, to jeśli już wiemy, że w naszym dobrym interesie leży skorzystanie z dobrodziejstw architektury Redux, to spójrzmy, co tak naprawdę przez to rozumiemy.
+
+![Redux](/assets/img/posts/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej/redux.png)
+<span class="img-legend">Architektura Redux z lotu ptaka. Źródło: dokumentacja NgRx</span>
+
+Na razie powyższy diagram może wydawać się nieco tajemniczy, zatem już spieszę z wyjaśnieniami.
+
+* Jak wspomniałem we wstępie, Redux zakłada istnienie globalnego, niemutowalnego bytu, przechowującego stan aplikacji - **store**. Fizycznie jest to obiekt w formacie JSON, którego struktura definiowana jest przez programistę. 
```

Comment:
Zawsze można błysnąć odnosząc się do CQRS. Dla backendowców ciekawych frontendu to może być taki moment "ooo, to wiem, to znam" ;)

Reply:
Ja i backend... :) Dodam, faktycznie może fajnie rozjaśnić